questionnaire_linking.py

```diff
@@ -0,0 +1,134 @@
+
+        if CASE_REF_ERROR_COUNT or QID_ERROR_COUNT:
+            print(f'Error: Line {line_number}: {questionnaire_link} has FAILED')
+        else:
+            post_message_to_queue(case_id, questionnaire_link["qid"], line_number)
+        print("\n")
+
+
+def validate_case_ref(line_number, case_ref):
+    global CASE_REF_ERROR_COUNT
+    if len(case_ref) != 8:
+        print(f'Error: Line {line_number}: Incorrect length of Case Reference {case_ref}. Expected length is 8 digits')
+        CASE_REF_ERROR_COUNT += 1
+    if not case_ref.isdigit():
+        print(f'Error: Line {line_number}: Incorrect type of Case Reference {case_ref}. Case Reference MUST be numeric')
+        CASE_REF_ERROR_COUNT += 1
+    if CASE_REF_ERROR_COUNT == 0:
+        response = requests.get(f'http://{Config.CASEAPI_HOST}:{Config.CASEAPI_PORT}/cases/ref/{case_ref}')
+        if response.status_code == 404:
+            print(f'Error: Line {line_number}: Case Reference {case_ref} does not exist')
+            CASE_REF_ERROR_COUNT += 1
+            return None
+        response.raise_for_status()
+        return response.json()["id"]
+
+
+def validate_qid_len_and_type(line_number, qid):
+    global QID_ERROR_COUNT
+    if len(qid) != 16:
+        print(f'Error: Line {line_number}: Incorrect length of QID {qid}. Expected length is 16 digits')
+        QID_ERROR_COUNT += 1
+    if not qid.isdigit():
+        print(f'Error: Line {line_number}: Incorrect type of QID {qid}. QID MUST be numeric')
+        QID_ERROR_COUNT += 1
+
+
+def validate_qid_third_digit(line_number, qid):
+    global QID_ERROR_COUNT
+    if len(qid) >= 3:
+        if qid[2] != "2":
+            print(f'Error: Line {line_number}: QID {qid} tranche identifier must be 2')
+            QID_ERROR_COUNT += 1
+    elif len(qid) < 3:
+        print(f'Error: Line {line_number}: QID {qid} must have a minimum of 3 digits for Tranche ID to be checked')
+
+
+def validate_check_digits(line_number, qid):
+    global QID_ERROR_COUNT
+    if len(qid) >= 3:
+        if not qid_checksum_validator.validate(qid, int(Config.QID_MODULUS), int(Config.QID_FACTOR))[0]:
+            print(f'Error: Line {line_number}: Check Digits incorrect in QID {qid}')
+            QID_ERROR_COUNT += 1
+    elif len(qid) < 3:
+        print(f'Error: Line {line_number}: QID {qid} length is not long enough for Check Digits to be validated')
+
+
+def post_message_to_queue(case_id, qid, line_number):
+    message = {
+        "event": {
+            "type": "QUESTIONNAIRE_LINKED",
+            "source": "RM",
+            "channel": "RM",
+            "dateTime": datetime.utcnow().isoformat() + 'Z',
+            "transactionId": str(uuid.uuid4())
+        },
+        "payload": {
+            "uac": {
+                "questionnaireId": qid,
+                "caseId": case_id,
+            }
+        }
+    }
+    with RabbitContext() as rabbit:
+        rabbit.publish_message(json.dumps(message), "application/json", None, exchange='events',
```

Comment:
We should probably send with the `delivery_mode` hardcoded to persistent here as we've recently highlighted that it doesn't necessarily default to this. 

You can do this by just making an pika `BasicProperties` object and setting the `delivery_mode`, then pass that into `publish_message` here as the properties, see https://github.com/ONSdigital/census-rm-print-file-service/pull/19